cord/ep: add proper user event notifications

haukepetersen opened the pull request:
### Contribution description
The solution I proposed in #12996 does not work properly for all cases and has its limitations. As alternative, this PR implements asynchronous event notifications for `cord_ep`, which in term are used and passed on by `cord_ep_standalone`. The code for `ep_standalone` should be a little simpler now, and if I am not mistaken there should be no duplicate event notifications anymore.

As further future, it is up to the user to select the actual events one wants be be notified about using the `mask` parameter int he `xyz_event_cb()` functions.

Caveats: the `xyz_event_cb()` functions are code duplicates right now, there is room for improvement...

**NOTE: some doxygen etc is still missing, will add this once the concept of this PR is approved...**

### Testing procedure
Run the endpoint example (`examples/cord_ep`) and register to a RD. Then play around with killing the RD (-> update should fail and the registration should be removed), unregistering manually (-> proper event notification should be displayed).

### Issues/PRs references
Alternative to #12996

miri64 commented:
@leandrolanzieri @pokgak @chrysn: @haukepetersen likes to have some feedback on this (https://github.com/RIOT-OS/RIOT/pull/12996#issuecomment-649589803).

pokgak commented:
Tested `cord_ep` example on native with [Californium `cf-rd`](https://github.com/eclipse/californium.tools/tree/master/cf-rd) as the RD. The application updates successfully. After killing the RD, the application will fail to update and after a while the registration is dropped. Manually unregister shows the notification as expected.

haukepetersen commented on sys/include/net/cord/ep.h:
```diff
@@ -106,11 +129,27 @@ int cord_ep_update(void);
+ *                      NULL to disable notifications
+ * @param[in] evt_mask
+ */
+void cord_ep_event_cb(cord_ep_cb_t cb, uint16_t evt_mask);
```
Yeah, I am fine with that, will change it.

haukepetersen commented:
rebased and removed the callback masking. We loose an insignificant bit of runtime performance, but gain API simplicity and simplified code, and a slight edge on ROM usage -> so seems to be worth it.

pokgak commented on sys/net/application_layer/cord/ep/cord_ep.c:
```diff
@@ -302,28 +311,26 @@ int cord_ep_register(const sock_udp_ep_t *remote, const char *regif)
     if (retval != CORD_EP_OK) {
         _rd_loc[0] = '\0';
```
We should also notify event `CORD_EP_DEREGISTERED` here.

Teufelchen1 commented:
See my comment in #12996
